File: 07 - Hash & Heap/E128hashFunction.py
class Solution:
    """
    @param key: A string you should hash
    @param HASH_SIZE: An integer
    @return: An integer
    """
    def hashCode(self, key, HASH_SIZE):
        # write your code here
        magicNumber, result = 33, 0

        for char in key:
            result = (result * magicNumber + ord(char)) % HASH_SIZE

        return result

File: 07 - Hash & Heap/E128hashFunction.java
public class Solution {
    /**
     * @param key: A string you should hash
     * @param HASH_SIZE: An integer
     * @return: An integer
     */
    public int hashCode(char[] key, int HASH_SIZE) {
        // write your code here
        int magic_code = 33;
        long result = 0;

        for (char ch: key) {
            result = (result * magic_code + ch) % HASH_SIZE;
            //if (result < 0) result += HASH_SIZE;
        }

        return (int) result;
    }
}
